sunxspex/thermal.py

```diff
@@ -0,0 +1,699 @@
+    somewhere along that same axis. That value does not have exactly correspond to
+    the value of a column in the grid. This is accounted for by the interpolation.
+
+    Parameters
+    ----------
+    logT: 1D `numpy.ndarray` of `float`.
+        The input value along the 2nd axis at which the line intensities are desired.
+        If multiple values given, the calculation is done for each and the
+        output array has an extra dimension.
+
+    data_grid: 2D `numpy.ndarray`
+        Some property, e.g. line intensity, as function two parameters,
+        e.g. energy (0th dimension) and log10(temperature in kelvin) (1st dimension).
+
+    line_logT_bins: 1D `numpy.ndarray`
+        The value along the 2nd axis at which the data are required,
+        say a value of log10(temperature in kelvin).
+
+    Returns
+    -------
+    interpolated_data: 1D or 2D `numpy.ndarray`
+        The line intensities as a function of energy (1st dimension) at
+        each of the input temperatures (0th dimension).
+        Note that unlike the input line intensity table, energy here is the 0th axis.
+        If there is only one input temperature, interpolated_data is 1D.
+
+    """
+    # Ensure input temperatures are in an array to consistent manipulation.
+    n_temperatures = len(logT)
+
+    # Get bins in which input temperatures belong.
+    temperature_bins = np.digitize(logT, line_logT_bins)-1
+
+    # For each input "temperature", interpolate the grid over the 2nd axis
+    # using the bins corresponding to the input "temperature" and the two neighboring bins.
+    # This will result in a function giving the data as a function of the 1st axis,
+    # say energy, at the input temperature to sub-temperature bin resolution.
+    interpolated_data = np.zeros((n_temperatures, data_grid.shape[0]))
+    for i in range(n_temperatures):
+        # Indentify the "temperature" bin to which the input "temperature"
+        # corresponds and its two nearest neighbors.
+        indx = temperature_bins[i]-1+np.arange(3)
+        # Interpolate the 2nd axis to produce a function that gives the data
+        # as a function of 1st axis, say energy, at a given value along the 2nd axis,
+        # say "temperature".
+        get_intensities_at_logT = interpolate.interp1d(line_logT_bins[indx], data_grid[:, indx], kind="quadratic")
+        # Use function to get interpolated_data as a function of the first axis at
+        # the input value along the 2nd axis,
+        # e.g. line intensities as a function of energy at a given temperature.
+        interpolated_data[i, :] = get_intensities_at_logT(logT[i]).squeeze()[:]
+
+    return interpolated_data
+
+
+def _weight_emission_bins_to_line_centroid(line_peaks_keV, energy_edges_keV, line_intensities):
+    """
+    Split emission between neighboring energy bins such that averaged energy is the line peak.
+
+    Given line peak energies and a set of the energy bin edges:
```

Comment:
this sentence is a little confusing maybe:

```suggestion
    Given the peak energies of the lines and a set of the energy bin edges:
```